camligorkem/assignment3/WordSearch.java

```diff
@@ -0,0 +1,123 @@
+import java.util.HashSet;
+import java.util.Optional;
+
+public class WordSearch {
+
+    /**
+     * Given a grid and a dictionary it finds the words (isWord=true) that exists in the dictionary
+     * and returns hashset of strings that are words
+     * @param dictionary is made up a list of entries which are either words or prefixes
+     * @param grid the board with the allowed letters
+     * @return return a hashset of strings that are found to be words (isWord = true)
+      */
+    public static Optional<HashSet<String>> findWords(Dictionary dictionary, char[][] grid ){
+
+        int n1 = grid.length;
+        int n2 = grid[0].length;
+        boolean[][] used;
+
+        // HashSet doesn't allow duplicate words, it is a good option to keep the found word
+        Optional<HashSet<String>> foundWords = Optional.empty();
+        Optional<HashSet<String>> res;
```

Comment:
I don't really see a reason to keep both of those variables. Why won't you just keep adding results of findWordsHelper runs to the foundWords set?